### PR TITLE
factorize advice task logic

### PR DESCRIPTION
**Pull request by stephanenicolas**
Extract the logic of advice task to prepare for transform task.

transform task will read the json file produced by advice and will either transform it and/or output it differently.

### CODE REVIEW

File: src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt [hunk 1]
```diff
@@ -131,61 +130,25 @@ abstract class AdviceTask : DefaultTask() {
...
-
-    if (!computedAdvice.filterAdd) {
-      advicePrinter.getAddAdvice()?.let {
-        consoleReport.add("Transitively used dependencies that should " +
-          "be declared directly as indicated:\n$it\n")
-        didGiveAdvice = true
-      }
-    }
-
-    if (!computedAdvice.filterChange) {
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt:155]: What happened to all these filters?

File: src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt [hunk 2]
```diff
@@ -2,15 +2,64 @@ package com.autonomousapps.internal.advice
 
 import com.autonomousapps.advice.Dependency
 import org.gradle.api.GradleException
+import org.jetbrains.annotations.TestOnly
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt:5]: Not really a fan of this. What is it needed?

File: src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt [hunk 3]
```diff
@@ -2,15 +2,64 @@ package com.autonomousapps.internal.advice
...
+    }
+    return ConsoleReport(consoleReport, didGiveAdvice)
+  }
+
   /**
    * Returns "add-advice" (or null if none) for printing to console.
    */
-  fun getAddAdvice(): String? {
+  @TestOnly fun getAddAdvice(): String? {
+    if(computedAdvice.filterAdd) {
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt:59]: Ah there it is.

File: src/main/kotlin/com/autonomousapps/internal/advice/ComputedAdvice.kt [hunk 4]
```diff
@@ -61,13 +61,12 @@ internal class ComputedAdvice(
   val unusedProcsAdvice: Set<Advice> = unusedProcs
     .mapToOrderedSet { Advice.remove(it.dependency) }
 
-  fun getAdvices(): Set<Advice> {
+  val advices: Set<Advice> by lazy {
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/internal/advice/ComputedAdvice.kt:64]: Assuming tests pass, I do quite like this change.

File: src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt (hunk 1 above)
Comment (autonomousapps) [reply src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt:155]: Nm, you internalized them.

File: src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt [hunk 5]
```diff
@@ -131,61 +130,25 @@ abstract class AdviceTask : DefaultTask() {
...
+    val consoleReport = advicePrinter.reportAdvicesToConsole()
 
-    if (didGiveAdvice) {
+    chatter.chat(consoleReport.toString())
+    if (consoleReport.hasAdvices) {
       chatter.chat("See machine-readable report at ${adviceFile.path}")
       chatter.chat("See pretty report at ${advicePrettyFile.path}")
       chatter.chat("See console report at ${adviceConsoleFile.path}")
-    } else {
-      consoleReport.add("Looking good! No changes needed")
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt:181]: You don't like this? :)

**Comment by stephanenicolas** [comment]
This PR is dropped in favor of an alternative approach : https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/pull/122

**Comment by autonomousapps** [comment]
Please rebase this onto master and force-push. I'm not using travis anymore, as it was timing out. CI is now done via Github Actions. (I have no confidence that travis's green light is legit.)

**Comment by autonomousapps** [comment]
Oh, so we can just close this PR?

**Comment by stephanenicolas** [comment]
Closed in favor of #122

**Pull request closed without merging**